Flat call tracer

ziogaschr opened the pull request.

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/internal/tracetest/testdata/call_tracer_flat
: add tests for flat call tracer

ziogaschr committed:
eth/tracers/native,go.mod,go.sum: add flatCall native tracer
:

ziogaschr committed:
basic formatting

ziogaschr committed:
add some undeclared structs

ziogaschr committed:
read callFrameMarshaling.TypeString in flatTracer

ziogaschr commented on eth/tracers/native/call_flat.go:
```diff
@@ -0,0 +1,447 @@
+		return nil, err
+	}
+
+	traceResult.Type = vm.StringToOp(traceResultMarshaled.TypeString)
```
Here we do this ugly hack to read, the **type** from the json.RawMessage passed from **callTracer**, where **type** is a string.
The thing is that if we don't find a better way to do that, this way has to be extended to do it recursively for the inners calls as well.

The root of it, is the Marshaling override [callFrameMarshaling](https://github.com/etclabscore/ethereum.go-ethereum/blob/457aba0f2ec22871ced3cbf0e0d85a2ca2581785/eth/tracers/native/call.go#L93), and next to it the generated code that is not defining `TypeString string         `json:"type"`` for the `UnmarshalJSON` [here](https://github.com/etclabscore/ethereum.go-ethereum/blob/457aba0f2ec22871ced3cbf0e0d85a2ca2581785/eth/tracers/native/gen_callframe_json.go#L64).

ziogaschr replied:
Resolved by https://github.com/etclabscore/ethereum.go-ethereum/pull/14/commits/271a8b897bf1650b5a51f423133e34aee1bc0d22

ziogaschr committed:
change callFrame.Type to stringOpCode type for gencodec marshaling

ziogaschr committed:
remove string quotes from stringOpCode.UnmarshalJSON

ziogaschr committed:
cleanup flatTracer code

ziogaschr committed:
remove debug code

ziogaschr committed:
remove dead code

ziogaschr committed:
go mod tidy

ziogaschr committed:
Merge branch 'master' into feat/flat-call-tracer

ziogaschr committed:
go mod tidy

ziogaschr committed:
fix flatCallTracer tests for using parity configuration (WiP)

ziogaschr committed:
flatCall tracer uses parent value for DELEGATECALL when zero

s1na commented on eth/tracers/native/call_flat.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+func (t *flatCallTracer) fillCallFrameFromContext(callFrame *flatCallFrame) {
+	if t.ctx != nil {
```
Hm I notice the context doesn't have the block number which is usually part of the parity trace. I'll have to look into why

ziogaschr replied:
**BlockNumber** has to be added, as said the PR is not finished.
I would like to know your thoughts on using a proxy tracers to `call_tracer` or a new tracer for flat_tracer.

The one thing I liked with the proxy one, is that we will have a single point in codebase to maintain the tracer logic. Although this approach, has some nitpicks in order to work.

Example nitpick which is not pushed here follows. Here and because we don't have access to `callTracer.callstack` we have to keep track of some stuff in this tracer ourselves. Making the `callTracer.**C**allstack` public will help, but then it will make more trivial to maintain the **callTracer** if we want to apply any changes.

```go
// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
func (t *flatCallTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
	// Update list of precompiles based on current block
	rules := env.ChainConfig().Rules(env.Context.BlockNumber, env.Context.Random != nil)
	t.activePrecompiles = vm.ActivePrecompiles(rules)

	// this is a workaround in order to exclude precompile calls in CaptureExit where we don't have the proxied tracer context
	t.currentCall = flatCallFrame{
		Type: strings.ToLower(vm.CALL.String()),
		Action: FlatCallTraceAction{
			From: &from,
			To:   &to,
			// Gas:   gas,
			// Value: value,
		},
	}

	if create {
		t.currentCall.Type = strings.ToLower(vm.CREATE.String())
	}

	t.tracer.CaptureStart(env, from, to, create, input, gas, value)
}
```

s1na replied:
> is that we will have a single point in codebase to maintain the tracer logic

I agree with that. As long as there aren't many work-arounds needed because of differences between parity-vs-geth styles like the one mentioned. That depends on whether we want a flatTracer or we want the parityTracer. Because as you mentioned parity tracer has some non-cosmetic differences like precompiles being excluded. Do you know what are the other differences?

You can cast the `t.tracer.(*callTracer)` and then you'll have access to `callstack` since they're in the same package. That should make things easier.

Comparing this version and the one in `core-geth` I must admit I like this one more. Even if we don't go with the proxy approach I think we should track the calls in the same shape that they will be returned (i.e. flat) so we don't need post-processing apart from toHex and toJSON conversions. I.e. avoid the `Finalize` method in your code. I realize a challenge is that we need access to the call in `CaptureExit`. But we can find the index of the call in the flat list by storing an additional stack that only tracks indices.

ziogaschr replied:
> Because as you mentioned parity tracer has some non-cosmetic differences like precompiles being excluded.

Do you find a bad idea to add an option in **CallTracer** that allows user to decide? `excludePrecompiles: boolean`

> You can cast the t.tracer.(*callTracer) and then you'll have access to callstack since they're in the same package. That should make things easier.

Nice one. Will check it and think if it adds more complexity than needed for maintenance later on.

s1na commented on eth/tracers/native/call_flat.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// GetResult returns an empty json object.
+func (t *flatCallTracer) GetResult() (json.RawMessage, error) {
```
Here we're doing an extra marshal and unmarshal. A call to `t.tracer.GetResult` is not necessary. We can directly access its callstack and process the output.

ziogaschr replied:
Yes, this make things a lot easier. Will try to work on this direction using this proxy tracer and see how it goes. 
Will be nice if you can reply here and with a go/no-go.

ziogaschr commented on eth/tracers/native/call.go:
```diff
@@ -89,13 +89,30 @@ func (f *callFrame) processOutput(output []byte, err error) {
-	Value      *hexutil.Big
-	Input      hexutil.Bytes
-	Output     hexutil.Bytes
+	Type    stringOpCode `json:"type"`
```
What do you think of these changes here @s1na? Was thinking that even this PR doesn't need this change, it is good to have it upstream for `gen_callframe_json.go`, so as if/when needed to be used, it will work as expected.

s1na committed:
fix top call frame gas used

s1na committed:
fix value inherited in delegatecall

s1na committed:
Exclude precompiles from calls

Co-authored-by: Chris Ziogas <ziogaschr@gmail.com>

s1na committed:
fix merge conflict

s1na committed:
copy addrs

s1na committed:
refactor types

s1na committed:
avoid json bounce in GetResult

s1na committed:
minor cleanups

ziogaschr committed:
s/convertedParityErrors/convertParityErrors for flat tracer tests

s1na committed:
fix test

s1na committed:
add blockNumber to context

s1na committed:
refactor result processing

ziogaschr committed:
eth/tracers/native: change struct fields to pointers to be ommited from JSON Marshaller

eth/tracers/native: change struct fields to pointers to be ommited from JSON Marshaller

ziogaschr committed:
eth/tracers/native: copy child call

ziogaschr committed:
eth/tracers/native: Child calls in flatCallTracer must have a value, even if it's zero

ziogaschr committed:
Merge branch 'master' into feat/flat-call-tracer

ziogaschr committed:
eth/tracers/native: update flatCallTracer to reflect new interfaces after merge

s1na committed:
fix lint issue

ziogaschr committed:
eth/tracers/internal/tracetest: update eth/tracers/internal/tracetest/flat_calltrace_test.go

Co-authored-by: Sina Mahmoodi <1591639+s1na@users.noreply.github.com>

ziogaschr committed:
eth/tracers/internal/tracetest: comment out debugging helpers

ziogaschr committed:
go.mod, go.sum: run `go mod tidy`

ziogaschr committed:
eth/tracers/native:  remove handling of value of a delegated call

This has been moved in callTracer context with PR https://github.com/ethereum/go-ethereum/pull/26632

ziogaschr committed:
eth/tracers/native: change call_tracer_flat tests config to follow the config schema of go-ethereum

ziogaschr committed:
Merge branch 'master' into feat/flat-call-tracer

ziogaschr committed:
eth/tracers/native: call_flat tracer will display calls to precompiles

For compatibility with Parity tracer we had removed the internal calls to precompiles for CALL/STATICCALL.

Geth decides to present them so as to present the full information in the tracers

ziogaschr committed:
eth/tracers/native: remove convertParityErrors from flat tracer

`convertParityErrors` tracer config option  has been used in order to keep errors same as in parity for compatibility.

Though, being compatibility in this level is not needed and is being removed for sanity

ziogaschr committed:
eth/tracers/native: remove unused code

ziogaschr committed:
eth/tracers/native: geth gas usage is more accurate, accounting for intrinsic gas and refunds

ziogaschr committed:
eth/tracers/native: flatCallTracer option to return Parity like formatted error for compatibility

ziogaschr committed:
eth/tracers/native: call_flat tracer option to handle calls to precompiles

For compatibility with Parity tracer we had removed the internal calls to precompiles for CALL/STATICCALL.

Geth decides to present them so as to present the full information in the tracers

For compatibility with Parity tracer we added the option `IncludePrecompiles` in order to include/exclude the internal calls to precompiles for CALL/STATICCALL.

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_flat: add test for “convertParityErrors” flatCall tracer option

s1na committed:
minor styling

ziogaschr committed:
eth/tracers/native: flatCall blockNumber and transactionPosition are non hex for compatibility

ziogaschr committed:
eth/tracers: fix panic by filling the BLockNumber for blocks and better handling nils in flatCall tracer

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_flat: fix tests for wrong gasUsed

s1na committed:
Revert gasUsed compatibility trick

s1na committed:
Revert gencodec type field conversion

s1na committed:
add comment

s1na committed:
Keep result in case of revert
